go-controller/go.mod

```diff
@@ -11,6 +11,7 @@ require (
 	github.com/containernetworking/cni v1.1.2
 	github.com/containernetworking/plugins v1.2.0
 	github.com/coreos/go-iptables v0.6.0
+	github.com/danwinship/knftables v0.0.12
```

Comment:
It's _pretty_ stable at this point. And no, the first user of it is kube-proxy (alpha nftables mode merged earlier this week) I just don't want to to declare it entirely stable yet. But if I make any changes to the API, I'm likely to be the one who has to update kube-proxy and ovn-k and openshift-sdn etc for them anyway. (Which means I'm probably not making any large incompatible changes to the API...)

There's a vague expectation that it will become API stable and get moved to https://github.com/kubernetes-sigs/ some time during the next kubernetes release cycle.

> why knftables uses the nft binary directly

Ah, I should explain this in the README. (We had discussed it in the KEP for kube-proxy...)

Basically, like with iptables, the only _documented_ nftables API is the API presented by the CLI tools, and the API at the netlink level is quite different, so if we used a low-level API, it would be a lot harder to understand the code and the rules that it was constructing (because they'd be built out of undocumented primitives and wouldn't look anything like the nftables rules in the man page or the wiki). And the CLI adds features that don't explicitly exist in the netlink API (e.g., IIRC, the low-level API doesn't explicitly have a concept of comments on rules/chains/etc. The CLI just has a particular convention for using certain metadata in certain ways.) I think it might have some backward-compat hacks for older kernels too...

That said, the CLI is actually a very thin wrapper around `libnftables.so`, which is where all the actual CLI API to netlink API conversion code lives, so in theory, knftables could link to that and call into it via cgo rather than execing the binary (without needing any changes to knftables's API). But that would probably be a lot more complicated to build/ship...

And anyway, ovn-k calls out to the iptables binaries to create iptables rules now, so calling out to the nftables binaries to do nftables rules doesn't put us in a _worse_ situation...

Reply:
> Ah, I should explain this in the README

https://github.com/danwinship/knftables/commit/51c65e8abf

Reply:
Sorry for the delay. I started looking at this as I am trying to fix an iptables issue in ovnk. What you said mostly makes sense, but it seems like google/nftables is a fairly well supported project and is used by a bunch of giant projects (although I'm not sure how many are just indirect imports). If they present us an API that  gives us what we need (Adding tables, chains, rules, etc - https://pkg.go.dev/github.com/google/nftables#New), then I think its worth considering.

Besides the community and more native netlink aspect, the other concern I have is with performance. Like I said the reason I started looking at this again was because I came across a bug where when a user has a large amount of services, our start up time is very long because we call iptables binary repeatedly for every single rule we have to add. I noticed in kubernetes-sigs/knftables there is support for creating transactions. Perhaps that would allow us to add all rules in a one shot operation?

Reply:
Following up on this some more: we've started running into more problems with the `nft` client in upstream k8s with version skew/bugginess, but these are going to need to be addressed in `nft` itself, because they also affect the ability of end users to run `nft list ruleset` from the command line to debug nftables.

But meanwhile, while investigating the idea of using a lower-level API, I started poking into the google nftables package, and looking at the open issues there, you repeatedly see things where the devs are telling people "yes, it seems like you ought to be able to do netlink commands 'X, Y, Z' to get the equivalent of `nft` commands 'x, y, z', but if you run `nft` with `--debug=netlink` you'll see that 'x, y, z' actually corresponds to 'X, A, 𝚿, Z', so that's what you actually have to do to make it work right".

So yeah, we don't want to be using the netlink API directly.

Reply:
I tend to agree here with @danwinship. https://github.com/google/nftables looks to be in the low API level closer to what `libnftnl` is, rather than a higher level API similar to `libnftables`. Unless we want to get into the business understanding and troubleshooting the netlink API (which we could, I don't think we miss the expertise, is just going to be extra time spent) to kinda come up with our higher level library ourselves, I think our best option might be `knftables` if we can afford the performance cost (which would be no detriment to what we have now for iptables) until something better pops up or we intentionally want to spend the extra time.

For example, in this issue https://github.com/google/nftables/issues/247, we have an example what would be an unexpected behavior when comparing to `nft` and the troubleshooting steps that we would have to do to resolve. I think it is reasonable to expect more instances of this happening.

On some occasions, the API documentation like https://pkg.go.dev/github.com/google/nftables@v0.2.0/expr#Match actually links links to the corresponding `libnftnl` code which in itself is not actually documented.

I couldn't find any thorough but easy to digest netlink api documentation either.

So I definitely think using https://pkg.go.dev/github.com/google/nftables is going to mean an extra time sink for us for a while rather than just something readily available for us to consume with no hiccups.

Reply:
After further discussion with @danwinship, it looks like we can consider `knftables` as that higher level API that we want to have because that is what is geared out to be as an API. Implementation wise, `knftables` may change over time, and use a lower level API in circumstances where it is required or would be beneficial, like resource intensive operations. From this perspective, it looks like an acceptable way forward. `knftables` would be the place where all of this lower level knowledge will be acquired over time instead of having to assume all that effort right now under the specific ovn-k umbrella.